_drafts/2020-09-17-fixing-an-n1-performance-problem-in-xaf-xpo-with-totally-undocumented-apis.md

```diff
@@ -0,0 +1,948 @@
+        ExecuteNonQueryCommand(@"CREATE VIEW [dbo].[FasterOfferWithViewQuery]
+AS select [FasterOfferWithView].[Oid] as Oid, 
+[FasterOfferWithView].[Oid] as [FasterOfferWithView], 
+[FasterOfferWithView].[Name] as [Name], 
+[FasterOfferWithView].[OptimisticLockField] as [OptimisticLockField], 
+[FasterOfferWithView].[GCRecord] AS [GCRecord],
+sum([FasterOfferItemWithView].[Hours]) AS [HourSum]
+from FasterOfferWithView
+inner join FasterOfferItemWithView on [FasterOfferWithView].[Oid] = [FasterOfferItemWithView].[FasterOfferWithView]
+group by [FasterOfferWithView].[Oid], [FasterOfferWithView].[Name], [FasterOfferWithView].[OptimisticLockField], [FasterOfferWithView].[GCRecord]
+", true);
+
+    }
+}
+
+public sealed partial class fixing_an_n_plus_1_perf_problem_in_xaf_xpoModule : ModuleBase
+{
+    //Register the updater in the module
+    public override IEnumerable<ModuleUpdater> GetModuleUpdaters(IObjectSpace objectSpace, Version versionFromDB) => new ModuleUpdater[]
+    {
+        new CreateViewDatabaseUpdater(objectSpace, versionFromDB)
+    };
+}
+
+```
+
+`Order`:
+
+```txt
+19.09.20 14:45:18.378 Executing sql 'select N0."Oid",N0."Name",N0."OptimisticLockField",N0."GCRecord" from "dbo"."FasterOfferWithView" N0 where N0."GCRecord" is null'
+19.09.20 14:45:18.380 Result: rowcount = 300, total = 12078, N0.{Oid,Guid} = 4800, N0.{Name,String} = 4878, N0.{OptimisticLockField,Int32} = 1200, N0.{GCRecord,Int32} = 1200
+19.09.20 14:45:18.388 Executing sql 'select top 1 count(*) from "dbo"."FasterOfferWithView" N0 where N0."GCRecord" is null'
+19.09.20 14:45:18.389 Result: rowcount = 1, total = 4, SubQuery(Count,,) = 4
+19.09.20 14:45:18.390 Executing sql 'select top 1 count(*) from "dbo"."FasterOfferItemWithView" N0 where N0."GCRecord" is null'
+19.09.20 14:45:18.410 Result: rowcount = 1, total = 4, SubQuery(Count,,) = 4
+```
+
+This results in the following performance
+
+1. Sort by `Name`: 0.009 seconds
+
+> *Hint*: we can still provide an `NonPersistent` `HourSum` field in the `Order` class, but make sure users can't display, filter, group by this field in any `ListView`. In `DetailViews` this should not be a huge performance penalty.
+
+`Query`:
+
+```txt
+19.09.20 14:46:07.290 Executing sql 'select top 2147483647 N0."HourSum",N0."Name",N0."Oid" from "dbo"."FasterOfferWithViewQuery" N0 where N0."GCRecord" is null'
+19.09.20 14:46:07.356 Result: rowcount = 300, total = 10878, N0.{HourSum,Int32} = 1200, N0.{Name,String} = 4878, N0.{Oid,Guid} = 4800
+19.09.20 14:46:07.360 Executing sql 'select top 1 count(*) from "dbo"."FasterOfferWithViewQuery" N0 where N0."GCRecord" is null'
+19.09.20 14:46:07.362 Result: rowcount = 1, total = 4, SubQuery(Count,,) = 4
+```
+
+This results in the following performance
+
+1. Sort by `Name`: 0.079 seconds
+2. Sort by `HourSum`: 0.070 seconds
+
+That's pretty impressiv! Finally let's have a look at the totally unsupported pitched version that saved me several times.
```

Comment:
```suggestion
That's very impressive! Finally let's have a look at the totally unsupported pitched version that saved me a lot of time.
```

Do you mean that you saved actual time: It saved me 10 hours or that it saved your life/ass many times?

Reply:
Both. 😂 But more the latter